File: leetcode/problems/problem-specific/1091.shortest-path-in-binary-matrix/1091.shortest-path-in-binary-matrix.go
package main

// * Iterative BFS Solution | O(mn) Time | O(mn) Space

func shortestPathBinaryMatrix(grid [][]int) int {
	if grid[0][0] == 1 {
		return -1
	}

	ROWS, COLS := len(grid), len(grid[0])
	queue := [][]int{}
	queue = append(queue, []int{0, 0, 1}) // [row, col, length]
	grid[0][0] = 1
	dir := [][]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}, {-1, -1}, {-1, 1}, {1, -1}, {1, 1}}
	for len(queue) > 0 {
		curSize := len(queue)
		for curSize > 0 {
			curCell := queue[0]
			queue = queue[1:]
			if curCell[0] == ROWS-1 && curCell[1] == COLS-1 {
				return curCell[2]
			}

			for _, v := range dir {
				row, col := curCell[0]+v[0], curCell[1]+v[1]
				if row >= 0 && row < ROWS && col >= 0 && col < COLS && grid[row][col] == 0 {
					queue = append(queue, []int{row, col, curCell[2] + 1})
					grid[row][col] = 1
				}
			}

			curSize--
		}
	}

	return -1
}

File: leetcode/problems/sessions/2023-05-01_2023-09-16/1091.shortest-path-in-binary-matrix.py
#
# @lc app=leetcode id=1091 lang=python3
#
# [1091] Shortest Path in Binary Matrix
#

# @lc code=start

# * Iterative BFS Solution | O(mn) Time | O(mn) Space
# * m -> The number of rows in the grid matrix | n -> The number of columns in the grid matrix


class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        DIRECTIONS = [
            (-1, 0),
            (1, 0),
            (0, -1),
            (0, 1),
            (-1, -1),
            (-1, 1),
            (1, -1),
            (1, 1),
        ]
        ROWS, COLS = len(grid), len(grid[0])
        if grid[0][0] == 1:
            return -1

        queue = deque([(1, 0, 0)])
        grid[0][0] = 1
        while queue:
            for _ in range(len(queue)):
                cur_len, cur_row, cur_col = queue.popleft()
                if cur_row == ROWS - 1 and cur_col == COLS - 1:
                    return cur_len

                for d_row, d_col in DIRECTIONS:
                    row, col = cur_row + d_row, cur_col + d_col
                    if (
                        not 0 <= row <= ROWS - 1
                        or not 0 <= col <= COLS - 1
                        or grid[row][col] == 1
                    ):
                        continue

                    queue.append((cur_len + 1, row, col))
                    grid[row][col] = 1

        return -1


# @lc code=end
